libraries/stdlib/samples/test/samples/coroutines/coroutines.kt

```diff
@@ -0,0 +1,146 @@
+        class FutureContinuation<T> : CompletableFuture<T>(), Continuation<T> {
+            override val context = EmptyCoroutineContext
+            override fun resumeWith(result: Result<T>) {
+                result.onSuccess { complete(it) }.onFailure { completeExceptionally(it) }
+            }
+        }
+
+        suspend fun <T> CompletableFuture<T>.await(): T =
+            suspendCoroutine { continuation ->
+                whenComplete { result, exception ->
+                    if (exception == null) // the future has been completed normally
+                        continuation.resume(result)
+                    else // the future has completed with an exception
+                        continuation.resumeWithException(exception)
+                }
+            }
+
+        val block: suspend () -> String = {
+            val bar = CompletableFuture.supplyAsync { "bar" }.await()
+            "foo$bar"
+        }
+        val result = FutureContinuation<String>().also { block.startCoroutine(it) }.join()
+        assertPrints(result, "foobar")
+    }
+
+    @Sample
+    fun codeFlow() {
+        // list to track the order of executed code
+        val list = mutableListOf<Int>()
+
+        // suspending lambda which stores its suspension point continuation in a local variable
+        var suspensionPoint: Continuation<Short>? = null
+        val block: suspend () -> Int = {
+            list += 2
+            val resumed = suspendCoroutine<Short> { suspensionPoint = it }
+            list += resumed.toInt()
+            5
+        }
+
+        // create a coroutine from the lambda, start and resume it after the suspension point
+        list += 0
+        val onComplete = Continuation<Int>(EmptyCoroutineContext) { list += it.getOrThrow() }
+        val init = block.createCoroutine(onComplete)
+        list += 1
+        init.resume(Unit) // starts the coroutine
+        list += 3
+        suspensionPoint!!.resume(4)
+        list += 6
+
+        assertPrints(list, "[0, 1, 2, 3, 4, 5, 6]")
+    }
+
+    @Sample
+    fun restrictsSuspension() {
+        @RestrictsSuspension
+        class SuspensionScope {
+            var result: String? = null
+            var continuation: Continuation<Unit>? = null
+            suspend fun suspend() {
+                suspendCoroutine<Unit> { continuation = it }
+                result = "foo"
+            }
+        }
+
+        val block: suspend SuspensionScope.() -> Unit = {
+            // this block can only invoke suspend functions on this receiver
+            // suspendCoroutine<Unit> {} // does no compile
+            this.suspend() // compiles
+        }
+
+        val receiver = SuspensionScope()
+        block.startCoroutine(receiver, Continuation(EmptyCoroutineContext) {})
+        receiver.continuation!!.resume(Unit)
+        assertPrints(receiver.result, "foo")
+    }
+
+    /**
+     * Unfortunately, this class cannot be a local class in a function because of the companion object.
```

Comment:
Unfortunately, this is a blocking obstacle for making this sample runnable in the documentation pages on kotlin website.

Reply:
I have a replacement for this sample which is now
* self-contained and
* also shows addition of context elements

The prototypical definition of a context element using `companion object` and `AbstractCoroutineContextElement` is now embedded in the kdoc of the latter.